lib/Dialect/SV/SVOps.cpp

```diff
@@ -817,23 +817,63 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
 }
 
 // If this wire is only written to, delete the wire and all writers.
-LogicalResult WireOp::canonicalize(WireOp op, PatternRewriter &rewriter) {
+LogicalResult WireOp::canonicalize(WireOp wire, PatternRewriter &rewriter) {
   // If the wire is named, then we can't delete it.
-  if (!op.name().empty())
+  if (wire.sym_nameAttr())
     return failure();
 
-  // Check that all operations on the wire are sv.connects. All other wire
-  // operations will have been handled by other canonicalization.
-  for (auto &use : op.getResult().getUses())
-    if (!isa<ConnectOp>(use.getOwner()))
+  // Wires have inout type, so they'll have connects and read_inout operations
+  // that work on them.  If anything unexpected is found then leave it alone.
+  SmallVector<sv::ReadInOutOp> reads;
+  sv::ConnectOp write;
+
+  for (auto *user : wire->getUsers()) {
+    if (auto read = dyn_cast<sv::ReadInOutOp>(user)) {
+      reads.push_back(read);
+      continue;
+    }
+
+    // Otherwise must be a connect, and we must not have seen a write yet.
+    auto connect = dyn_cast<sv::ConnectOp>(user);
+    // Either the wire has more than one write or another kind of Op (other than
+    // ConectOp and ReadInOutOp), then can't optimize.
+    if (!connect || write)
       return failure();
+    write = connect;
+  }
 
-  // Remove all uses of the wire.
-  for (auto &use : make_early_inc_range(op.getResult().getUses()))
-    rewriter.eraseOp(use.getOwner());
+  // If no write and only reads, then optimize away the wire.
+  if (!write) {
 
-  // Remove the wire.
-  rewriter.eraseOp(op);
+    bool canErase = true;
+    // Remove all uses of the wire.
+    for (auto &use : make_early_inc_range(wire.getResult().getUses())) {
```

Comment:
Please split the canonicalize patch out to its own separate patch, as it is logically distinct and the other patch already has a big effect on the testsuite.

When this comes back in, please change it to just RAUW the "readinout" ops with an X value.  That way you can always remove the wire when there is no writes.

Reply:
Thanks, created a separate PR for the canonicalize changes.